server/readonly/readonly_service.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  Logger,
+  OpenSearchDashboardsRequest,
+  SessionStorageFactory,
+} from '../../../../src/core/server';
+import { globalTenantName, isPrivateTenant } from '../../common';
+import { SecurityClient } from '../backend/opensearch_security_client';
+import { IAuthenticationType, OpenSearchAuthInfo } from '../auth/types/authentication_type';
+import { SecuritySessionCookie } from '../session/security_cookie';
+import { SecurityPluginConfigType } from '../index';
+import { ReadonlyService as BaseReadonlyService } from '../../../../src/core/server/security/readonly_service';
+
+export class ReadonlyService extends BaseReadonlyService {
+  private readonly logger: Logger;
+  private readonly securityClient: SecurityClient;
+  private readonly auth: IAuthenticationType;
+  private readonly securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>;
+  private readonly config: SecurityPluginConfigType;
+
+  constructor(
+    logger: Logger,
+    securityClient: SecurityClient,
+    auth: IAuthenticationType,
+    securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    config: SecurityPluginConfigType
+  ) {
+    super();
+    this.logger = logger;
+    this.securityClient = securityClient;
+    this.auth = auth;
+    this.securitySessionStorageFactory = securitySessionStorageFactory;
+    this.config = config;
+  }
+
+  isAnonymousPage(request: OpenSearchDashboardsRequest) {
+    if (!request.headers || !request.headers.referer) {
+      return false;
+    }
+
+    try {
+      const url = new URL(request.headers.referer as string);
+      const pathsToIgnore = ['login', 'logout', 'customerror'];
+      return pathsToIgnore.includes(url.pathname?.split('/').pop() || '');
+    } catch (error: any) {
+      this.logger.error(`Could not parse the referer for the capabilites: ${error.stack}`);
+    }
+  }
+
+  isReadOnlyTenant(authInfo: OpenSearchAuthInfo): boolean {
+    const currentTenant = authInfo.user_requested_tenant || globalTenantName;
+
+    // private tenant is not affected
+    if (isPrivateTenant(currentTenant)) {
+      return false;
+    }
+
+    return authInfo.tenants[currentTenant] !== true;
+  }
+
+  async isReadonly(request: OpenSearchDashboardsRequest): Promise<boolean> {
+    // omit for anonymous pages to avoid authentication errors
+    if (this.isAnonymousPage(request)) {
+      return false;
+    }
+
+    if (!this.config?.multitenancy.enabled) {
+      return false;
+    }
+
+    try {
+      const cookie = await this.securitySessionStorageFactory.asScoped(request).get();
+      let headers = request.headers;
+
+      if (!this.auth.requestIncludesAuthInfo(request) && cookie) {
+        headers = this.auth.buildAuthHeaderFromCookie(cookie, request);
+      }
+
+      const authInfo = await this.securityClient.authinfo(request, headers);
+
+      if (!authInfo.user_requested_tenant && cookie) {
+        authInfo.user_requested_tenant = cookie.tenant;
+      }
```

Comment:
I'm not fully familiar with these systems, could this be inside of the `authinfo(...)`, why have it here?

Reply:
@peternied I've experienced (at least at my setup) that `authInfo` (at this point) doesn't contain information about `requested_tenant`. Usually, it should be within `authInfo` but it looks like there are some cases where it's not.